fix: networkscenemanager not releasing buffers from pool

NoelStephensUnity opened the pull request:
This fixes the issue where the two SceneEventData instances in NetworkSceneManager were not putting their PooledNetworkBuffers back into the pool.  After further consideration, it seemed better to not hold onto PooledNetworkBuffers for the duration of the NetworkSceneManager's life time.  
Included in this fix:

-  SceneEventData creates one dedicated NetworkBuffer per instance
- NetworkSceneManager is now IDisposable
- NetworkSceneManager.Dispose now disposes both SceneEventData instances
- NetworkManager.Shutdown now disposes the NetworkSceneManager instance.

NoelStephensUnity committed:
fix

Don't used pooled NetworkBuffers for SceneEventData, instead just allocate 1 dedicated NetworkBuffer per SceneEventData (which will yield 2 instances per NetworkSceneManager).
Make sure we dispose the scene manager so it will dispose the two SceneEventData instances so they will Dispose their NetworkBuffers.

NoelStephensUnity committed:
style

Adding comments

NoelStephensUnity committed:
Merge branch 'develop' into fix/networkscenemanager-buffer-pool

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs:
```diff
@@ -85,7 +85,7 @@ public class SceneEvent
     /// Uses the <see cref="MessageQueueContainer.MessageType.SceneEvent"/> message to communicate <see cref="SceneEventData"/> between the server and client(s)
     /// </summary>
-    public class NetworkSceneManager
+    public class NetworkSceneManager:IDisposable
```
standards.py is going to be angry with you :)

NoelStephensUnity replied:
Fixed that one standards issue in most recent commit and fixed several other standards issues introduced when metrics were added.

jeffreyrainy commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/SceneEventData.cs:
```diff
@@ -754,7 +754,7 @@ public void Dispose()
         {
             m_NetworkManager = networkManager;
-            InternalBuffer = NetworkBufferPool.GetBuffer();
+            InternalBuffer = new NetworkBuffer(1024, 256);
```
How often is this triggered ?  Is this a worry for GC ?

NoelStephensUnity replied:
For every NetworkSceneManager instance there will be two (2) SceneEventData instances for the lifetime of that NetworkSceneManager instance.  There is always 1 NetworkSceneManager instance per network session (host, server, or client).
So, it will allocate roughly 2048 bytes total when StartHost, StartServer, or StartClient are invoked.  One of the two NetworkBuffers could increase in size during client synchronization (i.e. player is just approved and connecting to a session) depending upon how much has to be synchronized.

mattwalsh-unity replied:
I was following till I got here.  If we're cleaning ourselves up properly now, can we not go back to just using the PooledNetworkBuffer?  If not, can we have a comment saying why SceneEventData needs its own NetworkBuffer?

mattwalsh-unity replied:
Could you hoist what you just commented @NoelStephensUnity into comments in the code?

JesseOlmer replied:
I think it's fine to not use the pooled versions in this case (fine either way, honestly)... 2 instances created upon the construction of NetworkManager basically has the same impact as getting 2 from the pool since the pool doesn't pre-allocate any at all.

NoelStephensUnity replied:
Well, it was actual a bit lazy of me to do it that way.  This has since been fixed to make SceneEventData use only what it needs and then give the buffer back to the pool when no longer neeed.

mattwalsh-unity reviewed (changes requested):
Want to understand better why a NetworkBuffer is needed just for this class

JesseOlmer reviewed (approved):
lgtm

NoelStephensUnity committed:
refactor

Switching back to PooledNetworkBuffer.
Now it only gets a PooledNetworkBuffer when needed, and releases it back into the pool when it is no longer needed.

NoelStephensUnity committed:
Merge branch 'develop' into fix/networkscenemanager-buffer-pool

NoelStephensUnity committed:
style

Fixing one standards violation introduced in this PR and several more that were introduced when metrics were added.

NoelStephensUnity commented:
Now SceneEventData only gets a PooledNetworkBuffer when needed and places the PooledNetworkBuffer back into the pool when it is no longer needed.  So, we just saved 2048 bytes!

NoelStephensUnity committed:
style and refactor

Removed a LF that was not intended.
Added comments to SetInternalBuffer and ReleaseInternalBuffer while also making them both private as nothing else needs to be calling these two methods.

NoelStephensUnity committed:
refactor

Removing standards check.

NoelStephensUnity committed:
style

Removing 1 LF from SceneEventData constructor.
Slightly improved constructor description.
Removed 1 LF between ReleaseInternalBuffer and SetInternalBuffer
Removed 1 LF between AddSceneToSynchronize and SceneHandlesToSynchronize property.

NoelStephensUnity committed:
Merge branch 'develop' into fix/networkscenemanager-buffer-pool
